Wikipedia/Code/WMFZeroConfigurationManager.m

```diff
@@ -0,0 +1,166 @@
+ */
+- (void)fetchZeroConfigurationAndSetIsZeroRatedIfNecessary {
+    
+    // Note: don't nil out self.zeroConfiguration in this method
+    // because if we do we can't show its exit message strings!
+
+    //TODO: ensure thread safety so we can do this work off the main thread...
+    dispatch_async(dispatch_get_main_queue(), ^{
+        @weakify(self);
+        AnyPromise *promise = [AnyPromise promiseWithValue:nil];
+        promise = [self fetchZeroConfiguration].then(^(WMFZeroConfiguration *zeroConfiguration) {
+            @strongify(self);
+            
+            // If the config is not enabled its "message" will be nil, so if we detect a nil message
+            // set the isZeroRated to NO before we post the WMFZeroRatingChanged notification.
+            if(zeroConfiguration.message == nil){
+                self.isZeroRated = NO;
+                // Reminder: don't nil out self.zeroConfiguration here or the carrier's exit message won't be available.
+            }else{
+                self.zeroConfiguration = zeroConfiguration;
+                self.isZeroRated = YES;
+            }
+            
+        }).catch(^(NSError* error){
+            @strongify(self);
+            self.isZeroRated = NO;
+        });
+    });
+}
+
+- (AnyPromise *)fetchZeroConfiguration {
+    [self.zeroConfigurationFetcher cancelAllFetches];
+    return [self.zeroConfigurationFetcher fetchZeroConfigurationForSiteURL:[[[MWKLanguageLinkController sharedInstance] appLanguage] siteURL]];
+}
+
+- (void)updateZeroRatingAndZeroConfigurationForResponseHeadersIfNecessary:(NSDictionary *)headers {
+    NSAssert(headers != nil, @"Expecting response headers.");
+    if(!headers){
+        return;
+    }
+
+    bool zeroEnabled = self.isZeroRated;
+    
+    NSString* xCarrierFromHeader = [headers objectForKey:WMFZeroXCarrier];
+    bool hasZeroHeader = (xCarrierFromHeader != nil);
+    if (hasZeroHeader) {
+        NSString* xCarrierMetaFromHeader = [headers objectForKey:WMFZeroXCarrierMeta];
+        if ([self hasChangeHappenedToCarrier:xCarrierFromHeader orCarrierMeta:xCarrierMetaFromHeader]) {
+            self.previousPartnerXCarrier = xCarrierFromHeader;
+            self.previousPartnerXCarrierMeta = xCarrierMetaFromHeader;
+            [self fetchZeroConfigurationAndSetIsZeroRatedIfNecessary];
+        }
+    }else if(zeroEnabled) {
+        self.previousPartnerXCarrier = nil;
+        self.previousPartnerXCarrierMeta = nil;
+        self.isZeroRated = NO;
+    }
+}
+
+- (BOOL)hasChangeHappenedToCarrier:(NSString*)xCarrier orCarrierMeta:(NSString*)xCarrierMeta {
+    return !(
+             [self isNullableString:self.previousPartnerXCarrier equalToNullableString:xCarrier]
+             &&
+             [self isNullableString:self.previousPartnerXCarrierMeta equalToNullableString:xCarrierMeta]
+             );
+}
+
+- (BOOL)isNullableString:(nullable NSString*)stringOne equalToNullableString:(nullable NSString*)stringTwo {
```

Comment:
They were equivalent except in the nil case:

     NSString* s1 = nil;
     NSString* s2 = nil;

     [s1 isEqualToString:s2] // NO

     [self isNullableString:s1 equalToNullableString:s2] // YES

I suppose I could do the nil check to catch it else do 'isEqualToString:s2'...